README.md
# swift-hsdp-sdk

[![Swift](https://github.com/philips-software/swift-hsdp-sdk/actions/workflows/swift.yaml/badge.svg?branch=main)](https://github.com/philips-software/swift-hsdp-sdk/actions/workflows/swift.yaml)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

This SDK provides a simple interface to features offered by the IAM services of HSDP.

The output of the project is a libary that you can import as a package in your xcode environment.

## Table of contents

- [Using in your projects](#using-in-your-projects)
- [Basic usage](#basic-usage)
- [Supported APIs](#supported-apis)
- [Todo](#todo)
- [Issues](#issues)
- [Contact / Getting help](#contact--geting-help)
- [License](#license)


## Using in your projects

The library dependency can be included in your projects by adding https://github.com/philips-software/swift-hsdp-sdk as package collection.

## Basic Usage

```
import SwiftHsdpSdk

let hsdpSkd = SwiftHsdpSdk();

// Login Request
Task {
    let request = LoginRequest(username: "user@mail.com", password: "p@ssw0rd", 
    basicAuthentication: BasicAuthentication(username: "public-client", password: ""));

    let value = try await hsdpSkd.login(lr: request);
    let accessToken = value.access_token
    let refreshToken = value.refresh_token

}

// Introspect
Task {
    let request = IntrospectRequest(accessToken: accessToken, basicAuthentication: BasicAuthentication(username: "username",
    password: "p@ssw0rd"));

    let value = try await hsdpSkd.introspect(ir: request);
    let isTokenActive = String(value.active)
}

// Refresh Token

Task {
    let request = RefreshRequest(refreshToken: refreshToken, basicAuthentication: BasicAuthentication(username: "username",
    password: ""));

    let value = try await hsdpSkd.refresh(rr: request);
    let accessToken = value.access_token
}

// Revoke Token

Task {
    let request = RevokeRequest(token: accessToken, basicAuthentication: BasicAuthentication(username: "username", 
    password: "p@ssw0rd"));

    hsdpSkd.revoke(rr: request) {result in
        ""
    };
}



```

## Supported APIs

The current implementation covers only a subset of HSDP APIs. Additional functionality is built as needed.

- [x] IAM Identity and Access Management (IAM)
  - [ ] Access Management
    - [ ] Federation
    - [x] OAuth2
      - [x] OAuth2 Authorization
      - [x] OAuth2 Token Revocation
      - [ ] OpenID Connect UserInfo
      - [x] Introspect
      - [x] Session (refresh, terminate)
      - [ ] OpenID (configuration, JWKS)


Other services can follow later.


## Todo

- Implement more HSDP API calls


## Issues

- If you have an issue: report it on the [issue tracker](https://github.com/philips-software/swift-hsdp-sdk/issues)


## Contact / Getting help

Matthijs van Marion (<matthijs.van.marion@philips.com>) \
Martijn van Welie (<martijn.van.welie@philips.com>)

## License

See [LICENSE.md](LICENSE.md).

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) 
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [v0.1.0] - 2022-10-01
### Added
- Initial release with IAM OAuth2 functionality

Tests/SwiftHsdpSdkTest/SwiftHsdpSdkTest.swift
import XCTest
import Alamofire
import Mocker
@testable import SwiftHsdpSdk


final class SwiftHsdpSdkTest: XCTestCase {
    
    var hsdpSDK: IamOAuth2!

    override func setUp() {
        super.setUp()
        let configuration = URLSessionConfiguration.default
        configuration.protocolClasses = [MockingURLProtocol.self] + (configuration.protocolClasses ?? [])
        let sessionManager = Session(configuration: configuration)
        hsdpSDK = IamOAuth2(region: Region.EuWest, environment: .Prod, clientId: "test", clientSecret: "test", session: sessionManager);
    }
    
    override func tearDown() {
        super.tearDown()
    }
    
    func testLoginSucces() async throws {
        
        let apiEndpoint = URL(string: "https://iam-service.eu-west.philips-healthsuite.com/authorize/oauth2/token")!
        let mock = Mock(url: apiEndpoint, dataType: .json, statusCode: 200, data: [.post: MockedData.loginResponse.data])
        mock.register()
    
        let response = try await hsdpSDK.login(username: "username", password: "password");
        
        XCTAssertEqual(response.accessToken, "a5e488fc-7178-4cdw-937d-7136vg713a8d", "Access toking parsing is incorect")
    }
    
    func testLoginError() async throws {
        
        let apiEndpoint = URL(string: "https://iam-service.eu-west.philips-healthsuite.com/authorize/oauth2/token")!
        let mock = Mock(url: apiEndpoint, dataType: .json, statusCode: 401, data: [.post: MockedData.errorResponse.data])
        mock.register()
    
        do {
            let response = try await hsdpSDK.login(username: "username", password: "password");
         }
         catch let e as IamError {
             XCTAssertEqual(e, IamError.InvalidGrant)
         }
    }
    
}

public final class MockedData {
    public static let loginResponse: URL = Bundle.module.url(forResource: "loginResponse", withExtension: "json")!
    public static let introspectResponse: URL = Bundle.module.url(forResource: "introspectResponse", withExtension: "json")!
    public static let refreshResponse: URL = Bundle.module.url(forResource: "refreshResponse", withExtension: "json")!
    public static let errorResponse: URL = Bundle.module.url(forResource: "errorResponse", withExtension: "json")!
}

internal extension URL {
    var data: Data {
        return try! Data(contentsOf: self)
    }
}


Package.swift
// swift-tools-version: 5.6
// The swift-tools-version declares the minimum version of Swift required to build this package.

import PackageDescription

let package = Package(
    name: "swift-hsdp-sdk",
    platforms: [
        .macOS(.v12), .iOS(.v15)
    ],
    products: [
        // Products define the executables and libraries a package produces, and make them visible to other packages.
        .library(
            name: "SwiftHsdpSdk",
            targets: ["SwiftHsdpSdk"]),
    ],
    dependencies: [
        .package(url: "https://github.com/Alamofire/Alamofire.git", .upToNextMajor(from: "5.6.1")),
        .package(url: "https://github.com/WeTransfer/Mocker.git", .upToNextMajor(from: "2.3.0"))
    ],
    targets: [
        // Targets are the basic building blocks of a package. A target can define a module or a test suite.
        // Targets can depend on other targets in this package, and on products in packages this package depends on.
        .target(
            name: "SwiftHsdpSdk",
            dependencies: ["Alamofire"]),
        .testTarget(
            name: "SwiftHsdpSdkTest",
            dependencies: ["SwiftHsdpSdk","Mocker", "Alamofire"],
            resources: [.process("Resources")]),
    ]
)


Sources/SwiftHsdpSdk/IamError.swift

import Foundation

public enum IamError : String, Error {
    /// 404, path not found
    case PathNotFound = "404"
    
    /// Invalid grant, the credentials provided are not valid
    case InvalidGrant = "invalid_grant"
    
    /// The specified grant type is unsupported
    case UnsupportedGrantType = "unsupported_grant_type"
    
    /// The provided client failed to authenticate. The clientId or clientSecret are wrong
    case ClientAuthenticationFailed = "invalid_client"
    
    /// The request is invalid
    case InvalidRequest = "invalid_request"
    
    /// Token has expired
    case TokenExpired = "token_expired"
    
    ///
    case UnAuthorizedClient = "unauthorized_client"
    
    ///  An unexpected error occured
    case Other = "other"
}

public struct IamErrorResponse : Codable {
    public let error: String
    public let error_description: String
}


Sources/SwiftHsdpSdk/DomainModels.swift


import Foundation

public enum Environment : String {
    case Prod = "iam-service"
    case Dev = "iam-dev"
    case ClientTest = "iam-client-test"
}

public enum Region : String {
    case EuWest = "eu-west"
    case UsEast = "us-east"
    case UsWest = "us-west"

    public func getIamUrl(_ environment: Environment) -> String {
        return "https://\(environment.rawValue).\(rawValue).philips-healthsuite.com/"
    }
}

public struct LoginResponse : Codable {
    
    public let access_token : String
    public let refresh_token : String?
    public let scope : String
    public let token_type : String
    public let id_token : String?
    public let expires_in : UInt
}

public struct IntrospectResponse : Codable {
    
    public let active: Bool
    public let username: String?
    public let scope: String?
    public let exp: Int?
    let organizations: Organization?
}

public struct RefreshResponse : Codable {
    
    public let scope : String
    public let access_token : String
    public let token_type : String
    public let id_token : String?
    public let expires_in : String
}

struct Organization: Codable {
    
    let managingOrganization: String
    let organizationList: [OrganizationListItem]
}

struct OrganizationListItem: Codable {
    
    let organizationId : String
    let permissions : [String]
    
}

public struct Token : Codable {
    public let tokenType: String
    public let scopes: [String]
    public let accessToken: String
    public let refreshToken: String
    public let expiresIn: UInt
    public let timestamp: Date

    public init(tokenType: String = "", scopes: [String] = [], accessToken: String = "", refreshToken: String = "", expiresIn: UInt = 0) {
        self.tokenType = tokenType
        self.scopes = scopes
        self.accessToken = accessToken
        self.refreshToken = refreshToken
        self.expiresIn = expiresIn
        self.timestamp = .now
    }
    
    public init(from loginResponse: LoginResponse) {
        self.tokenType = loginResponse.token_type
        self.scopes = loginResponse.scope.components(separatedBy: " ")
        self.accessToken = loginResponse.access_token
        self.refreshToken = loginResponse.refresh_token ?? ""
        self.expiresIn = loginResponse.expires_in
        self.timestamp = .now
    }
    
    public func update(from refreshResponse: RefreshResponse) -> Token {
        return Token(
            tokenType: refreshResponse.token_type,
            scopes: refreshResponse.scope.components(separatedBy: " "),
            accessToken: refreshResponse.access_token,
            refreshToken: self.refreshToken,
            expiresIn: UInt(refreshResponse.expires_in) ?? 0
        )
    }
}


Sources/SwiftHsdpSdk/SwiftHsdpSdk.swift
import Foundation
import Alamofire
import SwiftUI


public class SwiftHsdpSdk {
    
    private let manager: Session
    private let url: HsdpUrlBuilder
    
    public init(manager: Session = Session.default, url: HsdpUrlBuilder = HsdpUrlBuilder()) {
        self.manager = manager
        self.url = url
    }
    
    public func getHsdpUrlBuilder() -> HsdpUrlBuilder {
        return url;
    }
    
    public func login(lr: LoginRequest) async throws -> LoginResponse
    {
        let basicAuthentication = "\(lr.basicAuthentication.username):\(lr.basicAuthentication.password)".data(using: .utf8)?.base64EncodedString()
        
        let headers: HTTPHeaders = [
            "Authorization": "Basic \(basicAuthentication!)",
            "Content-Type": "application/x-www-form-urlencoded",
            "API-version": "2",
            "Accept": "*/*"
        ]
        
        let parameters: Parameters = [
            "grant_type": "password",
            "username": "\(lr.username)",
            "password": "\(lr.password)"
        ]
        
        return try await manager.request("\(url.getIAMURL())\(url.tokenPath)", method: .post, parameters: parameters, headers: headers).serializingDecodable(LoginResponse.self).value
    }
    
    public func introspect(ir: IntrospectRequest) async throws -> IntrospectResponse
    {
        let basicAuthentication = "\(ir.basicAuthentication.username):\(ir.basicAuthentication.password)".data(using: .utf8)?.base64EncodedString()
        
        let headers: HTTPHeaders = [
            "Authorization": "Basic \(basicAuthentication!)",
            "Content-Type": "application/x-www-form-urlencoded",
            "API-version": "3",
            "Accept": "application/json"
        ]
        
        let parameters: Parameters = [
            "token": "\(ir.accessToken)"
        ]
        
        return try await manager.request("\(url.getIAMURL())\(url.introspectPath)", method: .post, parameters: parameters, headers: headers).serializingDecodable(IntrospectResponse.self).value
    }
    
    public func refresh(rr: RefreshRequest) async throws -> RefreshResponse
    {
        let basicAuthentication = "\(rr.basicAuthentication.username):\(rr.basicAuthentication.password)".data(using: .utf8)?.base64EncodedString()
        
        let headers: HTTPHeaders = [
            "Authorization": "Basic \(basicAuthentication!)",
            "Content-Type": "application/x-www-form-urlencoded",
            "Accept": "*/*"
        ]
        
        let parameters: Parameters = [
            "grant_type": "refresh_token",
            "refresh_token": "\(rr.refreshToken)"
        ]
        
        return try await manager.request("\(url.getIAMURL())\(url.tokenPath)", method: .post, parameters: parameters, headers: headers).serializingDecodable(RefreshResponse.self).value
    }
    
    public func revoke(rr: RevokeRequest, onCompletion: @escaping ( Optional<Int>) -> Void)
    {
        let basicAuthentication = "\(rr.basicAuthentication.username):\(rr.basicAuthentication.password)".data(using: .utf8)?.base64EncodedString()
        
        let headers: HTTPHeaders = [
            "Authorization": "Basic \(basicAuthentication!)",
            "Content-Type": "application/x-www-form-urlencoded",
            "Accept": "*/*",
            "API-version": "2"
        ]
        
        let parameters: Parameters = [
            "token": "\(rr.token)"
        ]
        
        manager.request("\(url.getIAMURL())\(url.revokePath)", method: .post, parameters: parameters, headers: headers).responseData(emptyResponseCodes: [200]) {
            data in
            onCompletion(data.response?.statusCode)
        }.resume()
        
    }
    
}


Sources/SwiftHsdpSdk/IamOAuth2.swift
import Foundation
import Alamofire

public class IamOAuth2 {
    
    public let session: Session
    public let baseURL : String
    public var token : Token = Token()
    public let basicAuthentication : String
    
    public let tokenPath = "authorize/oauth2/token"
    public let revokePath = "authorize/oauth2/revoke"
    public let introspectPath = "authorize/oauth2/introspect"
    public let userInfoPath = "authorize/oauth2/userinfo"
    

    public convenience init(region: Region, environment: Environment, clientId: String, clientSecret: String, token: Token = Token(), session: Session = Session.default) {
        self.init(baseURL: region.getIamUrl(environment), clientId: clientId, clientSecret: clientSecret, session: session)
    }
    
    public init(baseURL: String, clientId: String, clientSecret: String, token: Token = Token(), session: Session = Session.default) {
        self.baseURL = baseURL
        self.token = token
        self.session = session
        self.basicAuthentication = "\(clientId):\(clientSecret)".data(using: .utf8)!.base64EncodedString()
    }
    
    fileprivate func processResponse<T>(_ response: DataResponse<String, AFError>, onSucces: (String) throws -> T) throws -> T {
        let statusCode = response.response!.statusCode
        if case let .failure(error) = response.result {
            if (statusCode == 404) {
                throw IamError.PathNotFound
            } else {
                print(error)
                throw IamError.Other
            }
        }
        
        if (200...299 ~= statusCode ) {
            return try onSucces(response.value!)
        } else {
            let iamErrorResponse = try JSONDecoder().decode(IamErrorResponse.self, from: Data(response.value!.utf8))
            throw IamError(rawValue: iamErrorResponse.error) ?? .Other
        }
    }
    
    public func login(username: String, password: String) async throws -> Token
    {
        let headers: HTTPHeaders = [
            "Authorization": "Basic \(basicAuthentication)",
            "Content-Type": "application/x-www-form-urlencoded",
            "API-version": "2",
            "Accept": "*/*"
        ]
        
        let parameters: Parameters = [
            "grant_type": "password",
            "username": username,
            "password": password
        ]
        
        let response = await session.request(
            baseURL + tokenPath,
            method: .post,
            parameters: parameters,
            headers: headers
        ).serializingString().response
                
        return try processResponse(response) { responseString in
            let loginResponse : LoginResponse = try JSONDecoder().decode(LoginResponse.self, from: Data(response.value!.utf8))
            self.token = Token(from: loginResponse)
            return self.token
        }
    }
    
    public func introspect() async throws -> IntrospectResponse
    {
        return try await introspect(self.token)
    }
    
    public func introspect(_ token: Token) async throws -> IntrospectResponse
    {
        let headers: HTTPHeaders = [
            "Authorization": "Basic \(basicAuthentication)",
            "Content-Type": "application/x-www-form-urlencoded",
            "API-version": "3",
            "Accept": "application/json"
        ]
        
        let parameters: Parameters = [
            "token": token.accessToken
        ]
        
        let response = await session.request(
            baseURL + introspectPath,
            method: .post,
            parameters: parameters,
            headers: headers
        ).serializingString().response
        
        return try processResponse(response) { responseString in
            return try JSONDecoder().decode(IntrospectResponse.self, from: Data(response.value!.utf8))
        }
    }
    
    public func refresh() async throws -> Token
    {
        return try await refresh(self.token)
    }
    
    public func refresh(_ token: Token) async throws -> Token
    {
        let headers: HTTPHeaders = [
            "Authorization": "Basic \(basicAuthentication)",
            "Content-Type": "application/x-www-form-urlencoded",
            "Accept": "*/*"
        ]
        
        let parameters: Parameters = [
            "grant_type": "refresh_token",
            "refresh_token": token.refreshToken
        ]
        
        let response = await session.request(
            baseURL + tokenPath,
            method: .post,
            parameters: parameters,
            headers: headers).serializingString().response
        
        return try processResponse(response) { responseString in
            print(responseString)
            let refreshResponse : RefreshResponse = try JSONDecoder().decode(RefreshResponse.self, from: Data(response.value!.utf8))
            self.token = self.token.update(from: refreshResponse)
            return self.token
        }
    }
    
    public func revoke() async throws
    {
        return try await revoke(self.token)
    }
    
    public func revoke(_ token: Token) async throws
    {
        let headers: HTTPHeaders = [
            "Authorization": "Basic \(basicAuthentication)",
            "Content-Type": "application/x-www-form-urlencoded",
            "Accept": "*/*",
            "API-version": "2"
        ]
        
        let parameters: Parameters = [
            "token": token.accessToken
        ]
        
        let response = await session.request(
            baseURL + revokePath,
            method: .post,
            parameters: parameters,
            headers: headers
        ).serializingString(emptyResponseCodes: [200]).response
        
        try processResponse(response) { responseString in
                return
        }
    }
}



